src/Symfony/Component/Serializer/Serializer.php

```diff
@@ -221,12 +221,11 @@ private function getNormalizer($data, ?string $format, array $context)
                     continue;
                 }
 
-                if (!$normalizer instanceof CacheableSupportsMethodInterface) {
+                if (!$normalizer instanceof CacheableSupportsMethodInterface || !$normalizer->hasCacheableSupportsMethod()) {
                     $this->normalizerCache[$format][$type][$k] = false;
                 } elseif ($normalizer->supportsNormalization($data, $format)) {
                     $this->normalizerCache[$format][$type][$k] = true;
-
-                    return $normalizer;
+                    break;
```

Comment:
this was the bug spotted by @emodric: we cannot return here, as the non-cacheable normalizer weren't  called yet.